### DISCUSSION TITLE
Embed cpp tools completion into custom DSLs

### DISCUSSION THREAD

**Discussion by hildenc1** [microsoft/vscode-cpptools#8404] [Q&A]
We are currently trying to add vscode editing support to the eclipse eTrice project (https://www.eclipse.org/etrice/).
We got the intial highlighting and code completion for the "room" language to work using the lsp. Now we would like to add code completion to the action code blocks within the room language. Those blocks are written in c. We already tried to setup [Request-Forwarding](https://code.visualstudio.com/api/language-extensions/embedded-languages#request-forwarding) to use the existing c language support provided by the cpptools.

Unfortunately  vscode.executeCompletionItemProvider command does not return any suggestions. If we save the content of the virtual c file we are creating the saved file can be opened and the completion works as expected. But even if we pass the uri of the saved temporary file to the command we get an empty result.

I noticed that semantic tokens are provided for the standalone file but not for the embedded c code. I am not sure if that could be an issue. But, I could not figure out how to delegate semantic token creation (using provideDocumentSemanticTokens ) as I need the legend to register the provider (which I don't have for the cpptools).Is there any documentation / example on how to get code completion from the cpptools in an embedded language? Or has anyone done that successfully and is willing to share is path to success?

If not maybe someone has some more insights into what is needed for the cpp tools to produce completion suggestions?

**Comment by bobbrow** [comment]
I'm sorry, but we can't help with this.  The TypeScript code in our repository is open source with the MIT license, but the language server backing that code is closed source and proprietary.  The language server binaries we release with our extension are not licensed for use in products outside the Visual Studio family.

**Reply by hildenc1** [reply]
Hi @bobbrow,

thank you for your reply. I want to make sure I understand this correctly before we run into licence issues in the future.

My current understanding was that by using the request forwarding approach (linked in my original question) we delegate the production of code completion suggestions back to vscode which may then use any extension available. 
From my understanding this would also mean that the completion might actually be handled by any extension which provides suggestions for .c/.cpp files. If our extension is installed in a vscode instance which happens to have the cpptools installed and activated the suggestions would be provided by cpptools.
That way our extension would not include or directly use the cpptools language server binaries included with your extension.

We would like to give users of vscode a more or less seamless experience when mixing functionality implemented in c files with c detail code within the room language.

Please let me know if you would consider the request forwarding approach to be unlicences/unintended use.

**Reply by bobbrow** [reply]
Are you implementing a VS Code extension?  The link to your project page didn't have any reference to VS Code, so it looked like you were implementing your own IDE.  If you are a VS Code extension and are calling into our extension, that does not violate any terms of use and I apologize for the confusion.

If you are not a VS Code extension then I'm unfamiliar with how you could use the request forwarding to communicate back with VS Code. Is that what you're trying to do?

**Reply by hildenc1** [reply]
Thanks for the clarification :-).

I guess I was beein imprecise in my question which caused the confusion it the first place.

We are currently developing a vscode extension as an alternativ to the existing eclipse tooling. But, as it is currently in a very early alpha stage it did not make it on the project page, yet.

**Reply by bobbrow** [reply]
Thank you for clarifying.  What you are trying to do would be allowed, but what I think is going to block you is that cpptools does not currently operate on all URI types.  It only supports `file://`.  So if you are using the request forwarding and trying to invoke us via your middleware and it doesn't send us a file URI, we're not going to have any information about the virtual files you're operating with.

Related: #3343

**Comment by Colengms** [comment]
Hi @hildenc1 .  The legend for semantic tokens used by the C/C++ extension can be found in this repo.  https://github.com/microsoft/vscode-cpptools/blob/4b852288113f9c3c72b3d753c6f78acc711b3c38/Extension/src/LanguageServer/client.ts#L1005

Is it possible that the `executeCompletionItemProvider` is failing because the URI is not a 'file' URI, or does not refer to a file that is associated with the C or C++ language IDs?

**Reply by Rajsmit** [reply]
Hi @Colengms , I am also trying to get the code completions for CPP code fragments embedded in a custom DSL in a vscode extension using API 'vscode.executeCompletionItemProvider' as per setup mentioned in https://code.visualstudio.com/api/language-extensions/embedded-languages#request-forwarding. To achieve this we have a separate CPP file which contains all the CPP fragments The Uri is a valid 'file' URI and exists in the same workspace but in different folder.

The issue we are observing is the completion provider list from API 'vscode.executeCompletionItemProvider' is only returned correctly if the target CPP file is open in the editor and at least had focus once. if we close the CPP file the completion list is again having two template only #endregion and #region.

Could you please provide any help to make completion list work even if the file is not open in vscode editor?

**Reply by bobbrow** [reply]
@Rajsmit I don't think this is a behavior that we want to support. The lifetime of language services is tied to the "opened" state. If we break that then we will need some other state to tell us when to shut down a language server for a given translation unit and "shut down" is not something that we want to delegate to other extensions.

**Reply by Colengms** [reply]
Hi @Rajsmit .  I suspect some changes coming in 1.19.3 may actually enable your scenario.  This use case wasn't something we anticipated, so you may still run into issues.  But I'd suggest giving 1.19.3 a try, once it's available.

**Reply by Rajsmit** [reply]
Hi @bobbrow , Thanks for you information. As a the concept of request forwarding i was expecting the cpp language server will respond with completion list for the requested file. 

Hi @Colengms, That is a Great news. I am looking forward to test this feature in the next release 1.19.3

**Reply by bobbrow** [reply]
@Rajsmit I was wrong about needing another state to shut down a language server. We already prune unused servers when memory pressure accumulates which should address my concerns about orphaned processes.

**Reply by Rajsmit** [reply]
Hi @Colengms , 
I tested the feature again in version **v1.19.6** but the issue still persists. the code completions for CPP code fragment doesn't appear until the target .cpp file is opened and focused at least once. I am attaching the resulting code completions for both cases.

**CPP file is not opened:**
![withoutcppfileopen](https://github.com/microsoft/vscode-cpptools/assets/3623097/e233aa8a-516a-4ab1-831c-f9209b7d161f)

**CPP file is opened beside:**
![withcppfileopened](https://github.com/microsoft/vscode-cpptools/assets/3623097/48e0446f-7807-4c93-801c-cb555b2935dc)

Please suggest.

**Reply by Rajsmit** [reply]
Hi,

We have  found one solution which works after opening the CPP file in background using vscode API "vscode.workspace.openTextDocument" and it does trigger "textDocument/didOpen" on the file and then calling the code completion API. The code completion works only after first some edit in the cpp files.  for clangd extension it is working without doing any edit of cpp file in the IDE with the same approach. https://github.com/clangd/vscode-clangd/issues/589 . Could you please see if we have some option to achieve the same in C/C++ extension?

Thanks,
Rajesh Kumar.

**Reply by sean-mcmanus** [reply]
@Rajsmit I'm not aware of why completion would require an edit. If C_Cpp.loggingLevel is set to "Debug" do you see any logging that explains why completion isn't working. You could file a bug/issue on this.

Update: I created an issue at https://github.com/microsoft/vscode-cpptools/issues/12143

**Reply by Rajsmit** [reply]
Hi @sean-mcmanus , Thanks for creating the issue for the same. I collected the logs and feels like the it expects CPP files to be visible in IDE for doing parsing. i am adding log below for both scenario.

**With file opened in background**
LSP: (received) textDocument/didOpen: file:///d%3A/test2/pub-sub/rtistic-pub-doc/art-comp-test/tests/RIC_ChoicePointConditionCheck_redefined_1_target/Cap2.cpp
LSP: (received) cpptools/didChangeVisibleTextEditors
LSP: (invoked) textDocument/didOpen: file:///d%3A/test2/pub-sub/rtistic-pub-doc/art-comp-test/tests/RIC_ChoicePointConditionCheck_redefined_1_target/Cap2.cpp
LSP: (invoked) cpptools/didChangeVisibleTextEditors
LSP: (received) textDocument/completion: file:///d%3A/test2/pub-sub/rtistic-pub-doc/art-comp-test/tests/RIC_ChoicePointConditionCheck_redefined_1_target/Cap2.cpp (id: 69)
LSP: (invoked) textDocument/completion: file:///d%3A/test2/pub-sub/rtistic-pub-doc/art-comp-test/tests/RIC_ChoicePointConditionCheck_redefined_1_target/Cap2.cpp (id: 69)
Populating file name cache...
Done populating filename cache. Elapsed time: 28 ms
auto_complete::handle_completion: file:///d%3A/test2/pub-sub/rtistic-pub-doc/art-comp-test/tests/RIC_ChoicePointConditionCheck_redefined_1_target/Cap2.cpp (35:8)
Offering completion
LSP: Sending response (id: 69)


**With File opened in IDE**

LSP: (received) cpptools/didChangeVisibleTextEditors
LSP: (invoked) cpptools/didChangeVisibleTextEditors
LSP: (received) cpptools/didChangeVisibleTextEditors
LSP: (invoked) cpptools/didChangeVisibleTextEditors
LSP: (received) cpptools/didChangeActiveEditor: file:///d%3A/test2/pub-sub/rtistic-pub-doc/art-comp-test/tests/RIC_ChoicePointConditionCheck_redefined_1_target/Cap2.cpp
LSP: (received) cpptools/getCodeActions: file:///d%3A/test2/pub-sub/rtistic-pub-doc/art-comp-test/tests/RIC_ChoicePointConditionCheck_redefined_1_target/Cap2.cpp (id: 71)
LSP: (received) cpptools/resumeParsing
LSP: (received) cpptools/getDocumentSymbols: file:///d%3A/test2/pub-sub/rtistic-pub-doc/art-comp-test/tests/RIC_ChoicePointConditionCheck_redefined_1_target/Cap2.cpp (id: 72)
LSP: (invoked) cpptools/getDocumentSymbols: file:///d%3A/test2/pub-sub/rtistic-pub-doc/art-comp-test/tests/RIC_ChoicePointConditionCheck_redefined_1_target/Cap2.cpp (id: 72)
LSP: (received) cpptools/didChangeTextEditorSelection
Intellisense update pending for: file:///d%3A/test2/pub-sub/rtistic-pub-doc/art-comp-test/tests/RIC_ChoicePointConditionCheck_redefined_1_target/Cap2.cpp
LSP: (invoked) cpptools/didChangeActiveEditor: file:///d%3A/test2/pub-sub/rtistic-pub-doc/art-comp-test/tests/RIC_ChoicePointConditionCheck_redefined_1_target/Cap2.cpp
LSP: (invoked) cpptools/getCodeActions: file:///d%3A/test2/pub-sub/rtistic-pub-doc/art-comp-test/tests/RIC_ChoicePointConditionCheck_redefined_1_target/Cap2.cpp (id: 71)
LSP: Sending response (id: 71)
LSP: (invoked) cpptools/resumeParsing
LSP: (invoked) cpptools/didChangeTextEditorSelection
LSP: Sending response (id: 72)
IntelliSense update scheduled and TU acquisition started for: file:///d%3A/test2/pub-sub/rtistic-pub-doc/art-comp-test/tests/RIC_ChoicePointConditionCheck_redefined_1_target/Cap2.cpp
LSP: (received) cpptools/getDocumentSymbols: file:///d%3A/test2/pub-sub/rtistic-pub-doc/art-comp-test/tests/RIC_ChoicePointConditionCheck_redefined_1_target/Cap2.cpp (id: 73)
LSP: (invoked) cpptools/getDocumentSymbols: file:///d%3A/test2/pub-sub/rtistic-pub-doc/art-comp-test/tests/RIC_ChoicePointConditionCheck_redefined_1_target/Cap2.cpp (id: 73)
LSP: (received) cpptools/getCodeActions: file:///d%3A/test2/pub-sub/rtistic-pub-doc/art-comp-test/tests/RIC_ChoicePointConditionCheck_redefined_1_target/Cap2.cpp (id: 74)
LSP: (invoked) cpptools/getCodeActions: file:///d%3A/test2/pub-sub/rtistic-pub-doc/art-comp-test/tests/RIC_ChoicePointConditionCheck_redefined_1_target/Cap2.cpp (id: 74)
LSP: Sending response (id: 74)
LSP: (received) cpptools/getFoldingRanges: file:///d%3A/test2/pub-sub/rtistic-pub-doc/art-comp-test/tests/RIC_ChoicePointConditionCheck_redefined_1_target/Cap2.cpp (id: 75)
LSP: (invoked) cpptools/getFoldingRanges: file:///d%3A/test2/pub-sub/rtistic-pub-doc/art-comp-test/tests/RIC_ChoicePointConditionCheck_redefined_1_target/Cap2.cpp (id: 75)
LSP: (received) cpptools/getDocumentSymbols: file:///d%3A/test2/pub-sub/rtistic-pub-doc/art-comp-test/tests/RIC_ChoicePointConditionCheck_redefined_1_target/Cap2.cpp (id: 76)
LSP: (invoked) cpptools/getDocumentSymbols: file:///d%3A/test2/pub-sub/rtistic-pub-doc/art-comp-test/tests/RIC_ChoicePointConditionCheck_redefined_1_target/Cap2.cpp (id: 76)
LSP: Sending response (id: 75)
LSP: Sending response (id: 73)
LSP: Sending response (id: 76)
sending compilation args for D:\test2\pub-sub\rtistic-pub-doc\art-comp-test\tests\RIC_ChoicePointConditionCheck_redefined_1_target\Cap2.cpp
  include: D:\TEST2\PUB-SUB\RTISTIC-PUB-DOC\ART-COMP-TEST\TESTS\RIC_CHOICEPOINTCONDITIONCHECK_REDEFINED_1_TARGET
  include: C:\USERS\RAJESH_KUM\.VSCODE\EXTENSIONS\SECURE-DEV-OPS.CODE-REALTIME-1.0.3\TARGETRTS\INCLUDE
  include: C:\USERS\RAJESH_KUM\.VSCODE\EXTENSIONS\SECURE-DEV-OPS.CODE-REALTIME-1.0.3\TARGETRTS\TARGET\WINT
  include: C:\PROGRAM FILES (X86)\MICROSOFT VISUAL STUDIO\2017\BUILDTOOLS\VC\TOOLS\MSVC\14.16.27023\INCLUDE
  include: C:\PROGRAM FILES (X86)\WINDOWS KITS\10\INCLUDE\10.0.17763.0\UM
  include: C:\PROGRAM FILES (X86)\WINDOWS KITS\10\INCLUDE\10.0.17763.0\UCRT
  include: C:\PROGRAM FILES (X86)\WINDOWS KITS\10\INCLUDE\10.0.17763.0\SHARED
  include: C:\PROGRAM FILES (X86)\WINDOWS KITS\10\INCLUDE\10.0.17763.0\WINRT
  include: C:\PROGRAM FILES (X86)\WINDOWS KITS\10\INCLUDE\10.0.17763.0\CPPWINRT
  stdver: ms_c++17
  intelliSenseMode: windows-msvc-x64
LSP: (received) textDocument/hover: file:///d%3A/test2/pub-sub/rtistic-pub-doc/art-comp-test/tests/RIC_ChoicePointConditionCheck_redefined_1_target/Cap2.cpp (id: 77)
LSP: (invoked) textDocument/hover: file:///d%3A/test2/pub-sub/rtistic-pub-doc/art-comp-test/tests/RIC_ChoicePointConditionCheck_redefined_1_target/Cap2.cpp (id: 77)
LSP: Sending response (id: 77)
Tag Parser will be used for IntelliSense operations in: file:///d%3A/test2/pub-sub/rtistic-pub-doc/art-comp-test/tests/RIC_ChoicePointConditionCheck_redefined_1_target/Cap2.cpp
Update IntelliSense time (sec): 2.072
LSP: (received) cpptools/getCodeActions: file:///d%3A/test2/pub-sub/rtistic-pub-doc/art-comp-test/tests/RIC_ChoicePointConditionCheck_redefined_1_target/Cap2.cpp (id: 78)
LSP: (invoked) cpptools/getCodeActions: file:///d%3A/test2/pub-sub/rtistic-pub-doc/art-comp-test/tests/RIC_ChoicePointConditionCheck_redefined_1_target/Cap2.cpp (id: 78)
LSP: Sending response (id: 78)
LSP: (received) cpptools/getFoldingRanges: file:///d%3A/test2/pub-sub/rtistic-pub-doc/art-comp-test/tests/RIC_ChoicePointConditionCheck_redefined_1_target/Cap2.cpp (id: 79)
LSP: (invoked) cpptools/getFoldingRanges: file:///d%3A/test2/pub-sub/rtistic-pub-doc/art-comp-test/tests/RIC_ChoicePointConditionCheck_redefined_1_target/Cap2.cpp (id: 79)
LSP: Sending response (id: 79)
LSP: (received) cpptools/didChangeVisibleTextEditors
LSP: (invoked) cpptools/didChangeVisibleTextEditors
LSP: (received) cpptools/resumeParsing
LSP: (invoked) cpptools/resumeParsing
LSP: (received) cpptools/didChangeVisibleTextEditors
LSP: (invoked) cpptools/didChangeVisibleTextEditors
LSP: (received) cpptools/didChangeVisibleTextEditors
LSP: (invoked) cpptools/didChangeVisibleTextEditors
LSP: (received) cpptools/didChangeVisibleTextEditors
LSP: (invoked) cpptools/didChangeVisibleTextEditors
Intellisense update pending for: file:///d%3A/test2/pub-sub/rtistic-pub-doc/art-comp-test/tests/RIC_ChoicePointConditionCheck_redefined_1_target/Cap2.cpp
LSP: (received) cpptools/didChangeActiveEditor: file:///d%3A/test2/pub-sub/rtistic-pub-doc/art-comp-test/tests/RIC_ChoicePointConditionCheck_redefined_1_target/Cap2.cpp
LSP: (invoked) cpptools/didChangeActiveEditor: file:///d%3A/test2/pub-sub/rtistic-pub-doc/art-comp-test/tests/RIC_ChoicePointConditionCheck_redefined_1_target/Cap2.cpp
LSP: (received) cpptools/getCodeActions: file:///d%3A/test2/pub-sub/rtistic-pub-doc/art-comp-test/tests/RIC_ChoicePointConditionCheck_redefined_1_target/Cap2.cpp (id: 80)
LSP: (invoked) cpptools/getCodeActions: file:///d%3A/test2/pub-sub/rtistic-pub-doc/art-comp-test/tests/RIC_ChoicePointConditionCheck_redefined_1_target/Cap2.cpp (id: 80)
LSP: Sending response (id: 80)
LSP: (received) cpptools/didChangeTextEditorSelection
LSP: (invoked) cpptools/didChangeTextEditorSelection
LSP: (received) cpptools/resumeParsing
LSP: (invoked) cpptools/resumeParsing
LSP: (received) cpptools/didChangeTextEditorSelection
LSP: (invoked) cpptools/didChangeTextEditorSelection
IntelliSense update scheduled and TU acquisition started for: file:///d%3A/test2/pub-sub/rtistic-pub-doc/art-comp-test/tests/RIC_ChoicePointConditionCheck_redefined_1_target/Cap2.cpp
Update IntelliSense time (sec): 0.015
LSP: (received) cpptools/getFoldingRanges: file:///d%3A/test2/pub-sub/rtistic-pub-doc/art-comp-test/tests/RIC_ChoicePointConditionCheck_redefined_1_target/Cap2.cpp (id: 81)
LSP: (invoked) cpptools/getFoldingRanges: file:///d%3A/test2/pub-sub/rtistic-pub-doc/art-comp-test/tests/RIC_ChoicePointConditionCheck_redefined_1_target/Cap2.cpp (id: 81)
LSP: (received) cpptools/getDocumentSymbols: file:///d%3A/test2/pub-sub/rtistic-pub-doc/art-comp-test/tests/RIC_ChoicePointConditionCheck_redefined_1_target/Cap2.cpp (id: 82)
LSP: (invoked) cpptools/getDocumentSymbols: file:///d%3A/test2/pub-sub/rtistic-pub-doc/art-comp-test/tests/RIC_ChoicePointConditionCheck_redefined_1_target/Cap2.cpp (id: 82)
LSP: Sending response (id: 82)
LSP: Sending response (id: 81)
LSP: (received) cpptools/getCodeActions: file:///d%3A/test2/pub-sub/rtistic-pub-doc/art-comp-test/tests/RIC_ChoicePointConditionCheck_redefined_1_target/Cap2.cpp (id: 83)
LSP: (invoked) cpptools/getCodeActions: file:///d%3A/test2/pub-sub/rtistic-pub-doc/art-comp-test/tests/RIC_ChoicePointConditionCheck_redefined_1_target/Cap2.cpp (id: 83)
LSP: Sending response (id: 83)
LSP: (received) cpptools/resumeParsing
LSP: (invoked) cpptools/resumeParsing
LSP: (received) cpptools/didChangeVisibleTextEditors
LSP: (invoked) cpptools/didChangeVisibleTextEditors
LSP: (received) cpptools/didChangeVisibleTextEditors
LSP: (invoked) cpptools/didChangeVisibleTextEditors
LSP: (received) textDocument/completion: file:///d%3A/test2/pub-sub/rtistic-pub-doc/art-comp-test/tests/RIC_ChoicePointConditionCheck_redefined_1_target/Cap2.cpp (id: 84)
LSP: (invoked) textDocument/completion: file:///d%3A/test2/pub-sub/rtistic-pub-doc/art-comp-test/tests/RIC_ChoicePointConditionCheck_redefined_1_target/Cap2.cpp (id: 84)
auto_complete::handle_completion: file:///d%3A/test2/pub-sub/rtistic-pub-doc/art-comp-test/tests/RIC_ChoicePointConditionCheck_redefined_1_target/Cap2.cpp (35:8)
Offering completion
LSP: Sending response (id: 84)

Thanks,
Rajesh Kumar.

**Reply by sean-mcmanus** [reply]
@Rajsmit Fixed with https://github.com/microsoft/vscode-cpptools/releases/tag/v1.20.1

**Reply by Rajsmit** [reply]
Hi @sean-mcmanus , Tested with v1.20.1 (pre-release) now the code completion is working with first time itself without opening in IDE. Thanks for quick resolution.
